src/Server/ArrowFlightHandler.cpp

```diff
@@ -0,0 +1,348 @@
+    const arrow::flight::Ticket& request,
+    std::unique_ptr<arrow::flight::FlightDataStream>* stream)
+{
+    try
+    {
+        const std::string sql = request.ticket;
+
+        DB::ThreadStatus thread_status;
+        auto query_ctx = session->makeQueryContext();
+        query_ctx->setCurrentQueryId("arrow_flight");
+        CurrentThread::QueryScope query_scope(query_ctx);
+        setThreadName("ArrowFlight");
+
+        DB::QueryFlags flags;
+        auto [_, io] = DB::executeQuery(sql, query_ctx, flags, DB::QueryProcessingStage::Complete);
+        auto executor = std::make_unique<DB::PullingPipelineExecutor>(io.pipeline);
+
+        const DB::Block header = executor->getHeader();
+
+        DB::CHColumnToArrowColumn::Settings arrow_settings;
+        arrow_settings.output_string_as_string = true;
+
+        DB::CHColumnToArrowColumn converter(header, "Arrow", arrow_settings);
+
+        std::vector<DB::Chunk> chunks;
+        DB::Block block;
+
+        while (executor->pull(block))
+        {
+            chunks.emplace_back(DB::Chunk(block.getColumns(), block.rows()));
+        }
+
+        std::shared_ptr<arrow::Table> arrow_table;
+        converter.chChunkToArrowTable(arrow_table, chunks, header.columns());
+
+        auto maybe_combined = arrow_table->CombineChunks();
+        if (!maybe_combined.ok())
+        {
+            return arrow::Status::IOError("DoGet failed: cannot combine chunks: " + maybe_combined.status().ToString());
+        }
+
+        auto combined_table = maybe_combined.ValueOrDie();
+        arrow::TableBatchReader reader(combined_table);
+
+        std::vector<std::shared_ptr<arrow::RecordBatch>> batches;
+        std::shared_ptr<arrow::RecordBatch> batch;
+
+        while (true)
+        {
+            auto st = reader.ReadNext(&batch);
+            if (!st.ok())
+                return arrow::Status::IOError("DoGet failed while reading batch: " + st.ToString());
+
+            if (!batch)
+                break;
+
+            batches.emplace_back(std::move(batch));
+        }
+
+        if (batches.empty())
+            return arrow::Status::Invalid("DoGet failed: no data produced");
+
+        auto schema = batches.front()->schema();
+        auto reader_result = arrow::RecordBatchReader::Make(batches, schema);
+        if (!reader_result.ok())
+        {
+            return arrow::Status::IOError("DoGet failed: " + reader_result.status().ToString());
+        }
+
+        *stream = std::make_unique<arrow::flight::RecordBatchStream>(reader_result.ValueOrDie());
+        return arrow::Status::OK();
+    }
+    catch (const DB::Exception & e)
+    {
+        return arrow::Status::IOError("DoGet failed: " + e.displayText());
+    }
+}
+
+
+arrow::Status ArrowFlightHandler::DoPut(
+    const arrow::flight::ServerCallContext& /*context*/,
+    std::unique_ptr<arrow::flight::FlightMessageReader> reader,
+    std::unique_ptr<arrow::flight::FlightMetadataWriter> /*writer*/)
```

Comment:
Please add tests for reading and writing via this interface.

Reply:
added some tests